docs/architecture/adr-060-abci++-phase-i.md

```diff
@@ -0,0 +1,227 @@
+   ensure we do not discard transactions from the mempool in case `ProcessProposal`
+   fails.
+
+```go
+type PriorityMempool[T MempoolTx] struct {
+	queue   *PriorityQueue[MempoolTx]
+	senders map[string][]int64
+	// ...
+}
+```
+
+> The `PriorityMempool[T MempoolTx]` implementation will support Options such as
+> limiting the mempool size by a fixed number of bytes.
+
+Previous discussions<sup>1</sup> have come to the agreement that Tendermint will
+perform a request to the application, via `PrepareProposalRequest`, with a certain
+amount of transactions reaped from Tendermint's local mempool. The exact amount
+of transactions reaped will be determined by a local operator configuration.
+This is referred to as the "one-shot approach" seen in discussions.
+
+When Tendermint reaps transactions from the local mempool and sends them to the
+application via `PrepareProposalRequest`, the application will have to evaluate
+the transactions. Specifically, it will need to inform Tendermint if it should
+reject and or include each transaction. Note, the application can even _replace_
+transactions entirely with other transactions.
+
+When evaluating transactions from `PrepareProposalRequest`, the application will
+ignore _all_ transactions sent to it in the request and instead reap up to
+`PrepareProposalRequest.max_tx_bytes` from it's own mempool. There is no need to
+execute the transactions for validity as they have already passed CheckTx.
+
+### `ProcessProposal`
+
+The `ProcessProposal` ABCI method is relatively straightforward. It is responsible
+for ensuring validity of the proposed block containing transactions that were
+selected from the `PrepareProposal` step. In order to check validity of the proposed
+block, we must iterate over the list of transactions and execute them using the
+same mode/execution strategy as `CheckTx`. In other words, we execute each transaction
+using the AnteHandler only -- no messages are executed. However, we cannot just
+execute `CheckTx` again, because `BaseApp` already has a modified `checkState` at
+this point. So when executing `ProcessProposal`, we create a similar branched
+state, `processProposalState`, off of `deliverState`. Using `processProposalState`
+we execute the AnteHandler for each transaction. Note, the `processProposalState`
+is never committed and is completely discarded after `ProcessProposal` completes.
+
+We will only populate the `Status` field of the `ProcessProposalResponse` with
+`ACCEPT` if ALL the transactions were accepted as valid, otherwise we will
+populate with `REJECT`.
+
+### `DeliverTx`
+
+Since transactions are not truly removed from the app-side mempool during
+`PrepareProposal`, since `ProcessProposal` can fail and we do not want to lose
+transactions, we need to finally remove the transaction from the app-side mempool
+during `DeliverTx` since during this phase, the transactions are being included
+in the proposed block.
```

Comment:
Probably none of my business, but have you considered something like a TTL to make sure every transaction is eventually removed from the app-side mempool (e.g., to prevent corner-case TXs from filling up the mempool over time)?

Reply:
I'm not sure there is a direct need for a TTL as we already have a similar notion on txs directly and the mempool will be capped (by tx size).